beater/java_attacher/java_attacher.go

```diff
@@ -0,0 +1,155 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package javaattacher
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	"github.com/elastic/apm-server/beater/config"
+)
+
+type JavaAttacher struct {
+	cfg      config.JavaAttacherConfig
+	logLevel string
+	logger   *logp.Logger
+	logfn    logfn
+}
+
+type logfn func(...interface{})
+
+func New(cfg config.JavaAttacherConfig, logLevel string) (JavaAttacher, error) {
+	if cfg.JavaBin == "" {
+		if jh := os.Getenv("JAVA_HOME"); jh != "" {
+			cfg.JavaBin = filepath.Join(jh, "/bin/java")
+		} else {
+			bin, err := exec.LookPath("java")
+			if err != nil {
+				return JavaAttacher{}, fmt.Errorf("no java binary found: %v", err)
+			}
+			cfg.JavaBin = bin
+		}
+	} else {
+		// Ensure we're using the correct separators for the system
+		// running apm-server
+		cfg.JavaBin = filepath.FromSlash(cfg.JavaBin)
+	}
+	logger := logp.NewLogger("java-attacher")
+	return JavaAttacher{
+		cfg:      cfg,
+		logLevel: logLevel,
+		logger:   logger,
+		logfn:    setLogfn(logger, logLevel),
+	}, nil
+}
+
+// javaAttacher is bundled by the server
+// TODO: Figure out the real path
+var javaAttacher = filepath.FromSlash("/bin/apm-agent-attach-cli-1.24.0-slim.jar")
+
+func (j JavaAttacher) Run(ctx context.Context) error {
+	cmd := j.build(ctx)
+	j.logger.Debugf("starting java attacher with command: %s", strings.Join(cmd.Args, " "))
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+
+	donec := make(chan struct{})
+	defer close(donec)
+	go func() {
+		scanner := bufio.NewScanner(stdout)
+		b := struct {
+			Message string `json:"message"`
+		}{}
```

Comment:
```suggestion
		b := struct {
			Message string `json:"message"`
			LogLevel string `json:"log.level"`
		}{}
```

Reply:
I think we should probably also parse the timestamp and any additional fields. Perhaps another PR.